File: 3.x.x/crates-utopia/dns/src/query.rs
use crate::message::Message;
use crate::protocol::Protocol;

/// A DNS query together with its source. PHP `Utopia\DNS\Query`.
#[derive(Debug, Clone)]
pub struct Query {
    pub message: Message,
    pub ip: String,
    pub port: u16,
    pub protocol: Protocol,
}

impl Query {
    #[must_use]
    pub fn new(message: Message, ip: impl Into<String>, port: u16, protocol: Protocol) -> Self {
        Self {
            message,
            ip: ip.into(),
            port,
            protocol,
        }
    }
}

File: packages/dns/src/DNS/Query.php
<?php

declare(strict_types=1);

namespace Utopia\DNS;

/**
 * A DNS query together with its source: the decoded message, the client
 * address, and the transport protocol it arrived over.
 *
 * Resolver decorators use the source for cross-cutting concerns such as
 * tracing and rate limiting. Note that UDP source addresses are spoofable;
 * stream transports (TCP, HTTPS) carry verified peer addresses.
 */
final readonly class Query
{
    public function __construct(
        public Message $message,
        public string $ip,
        public int $port,
        public Protocol $protocol,
    ) {}
}
